test/e2e/upgrade_test.go

```diff
@@ -216,154 +230,164 @@ var _ = Describe("BMO Upgrade", Label("optional", "upgrade"), func() {
 			Expect(err).NotTo(HaveOccurred())
 		}
 	})
+	DescribeTable("Upgrade",
+		func(ctx context.Context, input *BMOIronicUpgradeInput) {
```

Comment:
This function is huge. It would be nice to define it separately just for better readability. A proper description of what exactly it is supposed to do is also necessary.
I know I have linked many times to CAPI and I will do it again here. Please check [how they describe](https://github.com/kubernetes-sigs/cluster-api/blob/5245bac52be028369285a0ee04205d073a6b864f/test/e2e/clusterctl_upgrade.go#L151-L180) the upgrade spec. I want to see something similar here. It is especially important to describe how/if there are differences between BMO and Ironic since the same function is used for both!

Reply:
Sure. I'll add some description! Thank you for pointing out the CAPI example!
I couldn't find any way to use an external func with `DescribeTable` (CAPI E2E has several `Describe` blocks calling the same function, which IMO is less nicer than `DescribeTable`). To be honest, I don't think that was an issue as it's currently also a function (without a name), but of course it would be nice. Would you have an example of how that could be done, which I could follow?

Reply:
It should be as simple as

```go
RunUpgradeTest(ctx context.Context, input *BMOIronicUpgradeInput) {
// ...
}

DescribeTable("Upgrade", RunUpgradeTest, ...)
```

Reply:
Yes, but some of variables used in that function (like the namespace or clusterProxy) are defined inside `BeforeEach`.

Reply:
Oh! Then that is the issue! We should not rely on variables from the context "outside" the function in that way. Either add what is needed to the BMOIronicUpgradeInput or move them inside the function.

Reply:
My bad, `namespace` is not defined in `BeforeEach`, but it and `cancelWatches` are defined inside the `RunUpgradeTest()`, and will be used later in `AfterEach()`.

Reply:
Adding this cleanup() inside `AfterEach` is kinda compulsory, because it should happen before BMO cleanup. If we add a `DeferCleanup()` right after namespace creation, BMO cleanup will happen first, as BMO upgrade happens after namespace creation. Otoh, if we put the namespace cleanup in a `DeferCleanup()` after BMO upgrade, it won't get run if something breaks the test before it gets to BMO upgrade :-?

Reply:
I may have found a solution. I will try to implement it.

Reply:
Function separated to its own place